content/automate/ManageIQ/Infrastructure/VM/Migrate/StateMachines/Methods.class/__methods__/checkmigration.rb

```diff
@@ -28,18 +28,27 @@ def checkmigration(task)
                 task_status = task['status']
                 result = task.statemachine_task_status
 
+                status = task.source.custom_get(:migration_status)
```

Comment:
The idea is to have a new `Vm#migration_status` method for all Vm types, basically RHEV and VMWare right now. For RHEV implementation you may use a customer attribute to track the status. 
You can set migration_status to `migrating` in the beginning, and `failed` or `succeeded` in the end.
But you don't directly get/delete customer attribute here.
You should not need  https://github.com/ManageIQ/manageiq-automation_engine/pull/126
and https://github.com/ManageIQ/manageiq-providers-ovirt/pull/168

For the new method `Vm#migration_status` you may want to return an array, first being the status, the second to be a detailed message explaining how it fails.

Reply:
The main goal for having [1] is to manage how state of `VmMigrateTask` is managed. At the moment what we need is different behavior for RHV so having generic behavior for everything and specific for RHV so [1] seems like good idea. 

Do you suggest to have additional column or use `CustomAttirbute` for which status is stored and retrieved by `Vm#migration_status`?

[1] https://github.com/ManageIQ/manageiq-providers-ovirt/pull/168

Reply:
We don't need [1] but have the initial status `active` for all providers, exactly the change you made before.
We will not add any additional column. The state machine retry (`checkprovision`) will keep on pulling `Vm#migration_status`. In its implementation you can read from custom attribute or directly query the event_streams table as we discussed earlier.

Reply:
This is how I understood your point about vmware and `vm.migrate` being done after returning from the call. With current changes vmware code would work and rhv would change the initial status to `active`.

Directly quering `event_streams` would mean that we need to depend on timestamps which is very error prone. We can move `custom_attribute` implementation to `vm` class.

Reply:
We want to change VMWare `vm.migrate` to return immediately and we will rely on `checkmigration` to wait for completion.

VMWare `vm.migrate` calls https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager.rb#L229
 and eventually https://github.com/ManageIQ/vmware_web_service/blob/9497671d9405c61fd3ab35603cab1b738a74986f/lib/VMwareWebService/MiqVimVm.rb#L196. Here we want to add `wait = true` in the argument list like other similar methods in this class. Make sure `vm.migrate` calls it with `wait = false`.

Then in `vm.migration_status` you can implement something similar to https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/provision/cloning.rb#L2

We need to think a way to pass around the `task_mor`, initially returned from `vm.migrate` method.

Reply:
OK so it is something different than I understood from our conversation. I will change.

Reply:
btw, how do you want to merge the code? Current changes (we need to add `vm.migration_status`) do not affect how other providers work. Do you want to wait on vmware migrate async code before merging this? I think it would be better to have this in and do not affect vmware migration and later when ready it could be merged.

Reply:
Adding `Vm#migration_status` in each provider can be merged first because they are new methods. Adding a flag (`wait = true`) in vmware_web_service can be merged early too.

Reply:
I can create `migration_status` in `VmOrTemplate` which should be visible to all the providers. I ask about migration state machine changes which will break vmware migration.

Reply:
state machine change should be merged last. We do not need to add `migration_status` immediately in `VmOrTemplate` because it will be an abstract method anyway. It is good to have but not required in runtime. Can be added later.

Reply:
why not to fix rhv issue and keep the current functionality for other providers and later work on vmware migration? Why do we need to split it?

Reply:
If you don't want to make changes to how VMWare migration works now, the minimum we can do is to implement VMWare's `migration_status` to always return `succeeded`.

Reply:
OK, I see where we misunderstood each other. I will go with `succeeded` for vmware.